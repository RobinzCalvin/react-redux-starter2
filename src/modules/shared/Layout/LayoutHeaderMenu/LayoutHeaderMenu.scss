@import '~shared/styles/mixins';

.layout-header-menu {
  $r: &;

  position: relative;
  display: flex;
  font-size: 0.9rem;

  &__title {
    display: none;

    @include breakpoint-up(sm) {
      display: inline;
    }
  }

  &__menu-icon {
    display: flex;

    @include breakpoint-up(sm) {
      display: none;
    }

    &_open {
      & ~ #{$r}__menu-items {
        position: absolute;
        top: 1.5rem;
        display: flex;
        flex-direction: column;
      }
    }
  }

  &__menu-items {
    display: none;

    @include breakpoint-up(sm) {
      display: flex;
      flex-direction: row;
      margin-left: 1rem;
    }
  }

  &__menu-item {
    width: 8rem;
    background: #444;
    color: #eaeaea;
    text-align: center;
    text-decoration: none;

    @include breakpoint-up(sm) {
      background: #333;
    }

    &_active {
      background: #666;

      @include breakpoint-up(sm) {
        background: #555;
      }
    }
  }
}
