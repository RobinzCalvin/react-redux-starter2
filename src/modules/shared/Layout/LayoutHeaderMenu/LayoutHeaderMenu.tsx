import React from 'react';
import block from 'bem-cn';
import { bind } from 'decko';
import { Link } from 'react-router-dom';

import { MenuIcon } from 'shared/view/elements';
import { ClickAwayListener } from 'shared/view/components';

import './LayoutHeaderMenu.scss';

interface IMenuItem {
  path: string;
  title: string;
}

interface IState {
  isMenuOpen: boolean;
}

interface IProps {
  menuItems: IMenuItem[];
}

const b = block('layout-header-menu');

class LayoutHeaderMenu extends React.PureComponent<IProps, IState> {
  public state: IState = {
    isMenuOpen: false,
  };

  public render() {
    const { isMenuOpen } = this.state;
    const { menuItems } = this.props;
    return (
      <div className={b()}>
        <div
          className={b('menu-icon', { open: isMenuOpen })}
          onClick={this.handleMenuClick}
          onTouchEnd={this.handleMenuTouchEnd}
        >
          <MenuIcon />
        </div>
        <span className={b('title')}>Search for:</span>
        <ClickAwayListener
          onClickAway={this.handleMenuItemsClickAway}
          mouseEvent="onClick"
          touchEvent="onTouchEnd"
        >
          <div className={b('menu-items')}>
            {menuItems.map(this.renderMenuItem)}
          </div>
        </ClickAwayListener>
      </div>
    );
  }

  private renderMenuItem({ path, title }: IMenuItem, i: number) {
    return (
      <Link to={path} className={b('menu-item', { active: path === location.pathname })} key={i}>
        {title}
      </Link>
    );
  }

  private toggleMenu(e: React.MouseEvent<HTMLDivElement> | React.TouchEvent<HTMLDivElement>) {
    e.preventDefault();
    this.setState((prevState: IState) => ({ isMenuOpen: !prevState.isMenuOpen }));
  }

  @bind
  private handleMenuClick(e: React.MouseEvent<HTMLDivElement>) {
    this.toggleMenu(e);
  }

  @bind
  private handleMenuTouchEnd(e: React.TouchEvent<HTMLDivElement>) {
    this.toggleMenu(e);
  }

  @bind
  private handleMenuItemsClickAway() {
    this.setState({ isMenuOpen: false });
  }
}

export { IMenuItem as IHeaderMenuItem, IProps as IHeaderMenuProps };
export default LayoutHeaderMenu;
