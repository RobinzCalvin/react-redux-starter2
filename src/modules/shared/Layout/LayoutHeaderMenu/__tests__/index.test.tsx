import React from 'react';
import { shallow } from 'enzyme';

import LayoutHeaderMenu, { IHeaderMenuProps } from '../LayoutHeaderMenu';

const props: IHeaderMenuProps = {
  menuItems: [
    {
      path: '/',
      title: 'Title',
    },
    {
      path: '/test',
      title: 'Title2',
    },
  ],
};

describe('LayoutHeaderMenu component', () => {
  it('should match snapshot', () => {
    const component = shallow(<LayoutHeaderMenu {...props} />);
    expect(component).toMatchSnapshot();
  });
});
