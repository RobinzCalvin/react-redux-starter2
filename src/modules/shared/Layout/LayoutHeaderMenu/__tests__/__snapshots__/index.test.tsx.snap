// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LayoutHeaderMenu component should match snapshot 1`] = `
<div
  className="layout-header-menu"
>
  <div
    className={
      Object {
        "has": [Function],
        "is": [Function],
        "mix": [Function],
        "split": [Function],
        "state": [Function],
        "toString": [Function],
      }
    }
    onClick={[Function]}
    onTouchEnd={[Function]}
  >
    <pure(MenuIcon) />
  </div>
  <span
    className={
      Object {
        "has": [Function],
        "is": [Function],
        "mix": [Function],
        "split": [Function],
        "state": [Function],
        "toString": [Function],
      }
    }
  >
    Search for:
  </span>
  <ClickAwayListener
    mouseEvent="onClick"
    onClickAway={[Function]}
    touchEvent="onTouchEnd"
  >
    <div
      className={
        Object {
          "has": [Function],
          "is": [Function],
          "mix": [Function],
          "split": [Function],
          "state": [Function],
          "toString": [Function],
        }
      }
    >
      <Link
        className={
          Object {
            "has": [Function],
            "is": [Function],
            "mix": [Function],
            "split": [Function],
            "state": [Function],
            "toString": [Function],
          }
        }
        key="0"
        replace={false}
        to="/"
      >
        Title
      </Link>
      <Link
        className={
          Object {
            "has": [Function],
            "is": [Function],
            "mix": [Function],
            "split": [Function],
            "state": [Function],
            "toString": [Function],
          }
        }
        key="1"
        replace={false}
        to="/test"
      >
        Title2
      </Link>
    </div>
  </ClickAwayListener>
</div>
`;
